executable-spec/src/Cardano/Ledger/Spec/STS/Update/Hupdate.hs

```diff
@@ -0,0 +1,112 @@
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DerivingVia #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+
+module Cardano.Ledger.Spec.STS.Update.Hupdate where
+
+import           GHC.Generics (Generic)
+--import           Data.Monoid.Generic (GenericMonoid (GenericMonoid),
+--                     GenericSemigroup (GenericSemigroup))
+import           Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+
+
+--import           Cardano.Crypto.Hash (Hash, HashAlgorithm)
+
+import           Control.State.Transition (Environment, PredicateFailure,
+                     STS, Signal, State, TRC (TRC), IRC (IRC), initialRules,
+                     judgmentContext, transitionRules)
+--import           Data.AbstractSize (HasTypeReps)
+
+import           Ledger.Core (Slot, BlockCount)
+import           Ledger.Core ((▷>=), (▷<=), (-.), (*.))
+
+
+-- import           Cardano.Ledger.Spec.STS.Update.Data ( SIPData
+--                                                      , Commit
+--                                                      )
+import qualified Cardano.Ledger.Spec.STS.Update.Data as Data
+
+
+
+-- | The Header Update STS
+-- Incorporates "update logic" processing
+-- at the block header level
+data HUPDATE hashAlgo
+
+data Env
+ = Env { k :: !BlockCount
+         -- ^ Chain stability parameter.
+       , currentSlot :: !Slot
+       }
+       deriving (Eq, Show)
+
+data St hashAlgo
+  = St { wrsips :: !(Map (Data.SIPHash hashAlgo) Slot)
+       -- ^ When a SIP was revealed
+       , asips :: !(Map (Data.SIPHash hashAlgo) Slot)
+       -- ^ When a SIP will not be active any more
+       -- (i.e., end of open for voting period)
+       }
+       deriving (Eq, Show, Generic)
+       -- deriving Semigroup via GenericSemigroup (St hashAlgo)
+       -- deriving Monoid via GenericMonoid (St hashAlgo)
+
+-- deriving instance Generic (Slot)
+-- deriving instance Semigroup (Slot)
+-- deriving instance Monoid (Slot)
+
+instance  STS (HUPDATE hashAlgo) where
+
+  type Environment (HUPDATE hashAlgo) = Env
+
+  type State (HUPDATE hashAlgo) = St hashAlgo
+
+  type Signal (HUPDATE hashAlgo) = Slot
+
+  data PredicateFailure (HUPDATE hashAlgo)
+    = ErrorOnHUpdate Slot Slot
+    deriving (Eq, Show)
+
+
+  initialRules = [
+    do
+      IRC Env { } <- judgmentContext
+      pure $! St { wrsips = Map.empty
+                 , asips = Map.empty
+                 }
+    ]
+
+  transitionRules = [
+    do
+      TRC ( Env { k }
+          , St  { wrsips
+                , asips
+                }
+          , slot
+          ) <- judgmentContext
+
+      let -- exclude old revealed SIPs
+          wrsips' = wrsips ▷>= (slot -. (2 *. k))
```

Comment:
It seems we might have an overlap in the two restrictions: namely we'll be keeping `slot - 2k` in `wrsips'` and `asips'`. Maybe using domain restriction would be better.